### PR TITLE
If we have an annotation, probably the Author means it.

### LINKED ISSUE #45173: SSL termination at ELB and internal communication over http on port 80
This is what I would like to setup

**Would like to have**
External access `only` via https on 443 and ssl terminated at ELB and all the requests get routed to the internal services via http on port 80. Also, all the internal services communications are all over http port 80.

**What's happening right now**
With the following setup, it results in ELB with https on port 443 and aws cert setup correctly. Internal ports are all mapped to port 443 as well, internal services cannot communicate via port 80.

**Work around**
current workaround is to create similar setup for port 80 and manually change the ELB to https port 443.

k8s service definition 
```
kind: Service
apiVersion: v1
metadata:
  name: test-app
  labels:
    name: test-app
    environment: dev
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-ssl-cert: <cert arn>
    service.beta.kubernetes.io/aws-load-balancer-backend-protocol: http
spec:
  type: "LoadBalancer"
  ports:
    - name: "https"
      port: 443
      targetPort: 80
  selector:
    name: ecommerce
    environment: dev
```

**Kubernetes version** : 1.4.x, 1.5.x

**Environment**:
- **Cloud provider**: AWS
- **OS** : Debian GNU/Linux 8 (jessie)

### PR DESCRIPTION
**Pull request by iMartyn**
The simple check to see if there is a numbered port is probably enough
to verify we are passing a number to AWS.

**What this PR does / why we need it**:
Currently if you specify an SSL ELB Port and your service listens on something other than that port (for example, your http (unencrypted) service listens on port 80, the ELB created is actually not SSL'd and listens on the port of the service (in this example port 80).  To maintain the logical entity of the Service as internally used as separate from the LoadBalancer that is externally used, the SSL ELB Port should override the assumption that we want the LoadBalancer to be exactly the same as the Service.
More notes on #45173 but we shouldn't force internal services to change for the purposes of external cloud-provided LoadBalancers, that should be transparent.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #45173 

**Special notes for your reviewer**:

**Release note**:

```release-note
If the user has requested an SSL ELB port number, we now respect that request and pass it to AWS rather than expecting the Port of the service to be used.
```

### CODE REVIEW

**Comment by iMartyn** [comment]
So, this fix breaks the test case "Port not in whitelist, passthrough" but I do not think that's good behaviour (listen on 443 if port is named and not in a whitelist).
I do still believe this to be wrong behaviour and a bug, but without others on board (facing a lot of "this isn't a bug") I don't know enough to decide between _change the test to match the behaviour_ or _figure out some other code gymnastics to use the requested port_.

**Comment by iMartyn** [comment]
Due to lack of comments/interaction on this issue, I have fixed the test to match the new behaviour.
This will make this mergeable and if it ever gets any traction is then at least complete.

**Comment by jsafrane** [comment]
@kubernetes/sig-aws-misc

**Comment by jsafrane** [comment]
@nckturner @micahhausler, PTAL, I am more storage-related, not familiar with networking.

BTW, did you get notification from kubernetes/sig-aws-misc one or two messages above? It would be useful if you're in these groups so we don't need to tag you individually.

**Comment by jsafrane** [comment]
@iMartyn, please sign CLA, see https://github.com/kubernetes/kubernetes/pull/56085#issuecomment-345795594

**Comment by iMartyn** [comment]
@jsafrane Unfortunately the cla-bot is fundementally broken.  As you can see, I did sign the cla but now I will need to rebase all the commits to update the email as it has unsigned all my previous commits.

**Comment by iMartyn** [comment]
I think you understand but as I want to be sure, let me storify it to make it easier to understand.

1. I make a new microservice that will deal with say profile information.
1. I deploy it as a pod and expose it as profile-service as a ClusterIP.
1. I set other microservices to use it - http://profile-service/
1. Everythings looking good, I need to expose it.  Because it's going to be used externally, better put https on it.  Oh look, ELBs will take care of that for me.
1. Add cert in AWS and give it's arn to the service and upgrade it from ClusterIP to LoadBalancer.  Whilst I'm at it, I'll give it a dns name.
1. Visit https://profile-service.mycoolapp.com/ no dice.  Service needs to be listening on port 443 to do that.  But the SSL is terminated at the ELB, https->http so why? "Because."

At this point I have many wrong things I can do to work round the issue :

- I can revert the upgrade to LoadBalancer and expose the service as something like profile-service-external listening unencrypted on port 443 with `type: Loadbalancer` and DNS name.  - Wrongness : 
a. I shouldn't have to have two services pointing at the same set of pods
b. This creates (even if not used) http://profile-service-external:443/ (note not https)
- I can add 443 to the existing service as well as 80 - Wrongness:
a. This creates (even if not used) http://profile-service:443 (note not https) and I have to explain to random people why https://profile-service/ doesn't work when it's clearly listening on port 443 - `kubectl get svc` says so!
- I can move the port to 443 and reconfigure everything to point to http://profile-service:443/ because sod it, we are being surprising, why not just embrace it.
- Terminate https within the clluster which when it can be terminated at the edge is not what I want to do.
- Probably other bizzare things but they all result in confusing or surprising setups.

> The way I've always understood it is that services with type: "LoadBalancer" are for configuring the cloud load balancer, but if internal services want plain HTTP access, creating a second service (like @justinsb suggests) with the intuitive and expected ports solves the issue.

I respectfully but fundamentally disagree with this, and I think this is the core of the issue.  I'll explain how I view services and perhaps it will make sense why I disagree so vehemently :
- Every service that is created in Kubernetes gets a dns name and a port for use within the cluster. `type: ClusterIP` does this.
- To access those services externally you can upgrade this to a `type: NodePort` and you will get a high port to access the service from.
- If your topology is such that you have a programmable load balancer, you can upgrade NodePort to `type: LoadBalancer`.  This still creates a NodePort and still creates a ClusterIP.

To me there is a clear building of ClusterIP + NodePort + LoadBalancer that is a progression, and the configuration of the LoadBalancer is done via annotations (I mean, why else is there a port annotation there?!).  Therefore you should absolutely not have to change the configuration of the underlying service main constructs to configure your LoadBalancer, and especially not forcing you to run on a port you don't want to run on (because it makes no sense and confuses people).

**Comment by micahhausler** [comment]
@iMartyn You have a convincing argument. I am personally ok with this, but I'd want the input of sig-network on this.

/sig network

**Comment by iMartyn** [comment]
@justinsb please can you weigh in, I'm at frustration point and I don't even use AWS at work anymore.

**Comment by nckturner** [comment]
@iMartyn I think I understand this change, basically you want the service to remain with:
```
spec:
  type: "LoadBalancer"
  ports:
    - name: "http"
      port: 80
      targetPort: 80
```
while allowing an LB to be created that listens on 443 with a target port of 80?  I probably have a simplified understanding of it, but the goal of keeping the LB from being tied to the service seems like a good thing.  I want to spend some time to test it but initially I think I agree with this change.  Sorry its been sitting around for so long, maybe we can get it merged...

**Comment by iMartyn** [comment]
@nckturner Yes, the annotations that are AWS specific should deal with the creation of the LB port on 443 without creating a service listening in-cluster on 443.

Fair warning, I no longer work with AWS so whilst I'd love to know this is fixed (especially if it's my code that gets merged), I can't actually test it any more.

**Comment by iMartyn** [comment]
/remove-lifecycle stale
Continues screaming into the void.

**Comment by iMartyn** [comment]
I give up, seriously, @justinsb @nckturner @micahhausler @jsafrane I really wanted to improve things here on something that I feel is fundamentally wrong behavior, but noone is listening who is in a position to accept this contribution.

**Comment by nikopen** [comment]
@dims @mcrute This seems to have been abandonded by the original OP, do you intend on continuing to work here or in another PR? Code Freeze is starting this Friday, a general deadline would be 10 days from now for 1.14.

**Comment by mcrute** [comment]
@nikopen let's try to get this into the merge window for 1.14. I'll do some testing this week and if everything works as expected we should be able to merge before Friday.

**Comment by iMartyn** [comment]
I think "abandoned" is a little unfair, I only abandoned hope that it would ever get picked up!

If you need extra input from me I'm happy to help, just bear in mind I don't work with aws any more so testing in a real environment is rather "tricky" now.

**Comment by claurence** [comment]
/hold

1.14 enhancement lead here - we've passed code freeze and I see this issue is still open and it sounds like some testing was going to happen this week - do we feel confident we can include this in 1.14?

**Comment by mcrute** [comment]
@claurence I was able to do some testing but not enough to be comfortable making the 1.14 freeze. Let's tag this for 1.15 and I'll push it forward.

@iMartyn thanks for your contribution, I think this is worth pushing forward. The whole way that SSL protocol selection for ELB is done strikes me as a little dodgy and I just want to validate my understanding and run through a few more test cases to make sure that this is the right technical approach. I think we should be able to land this fix or something very similar to it for 1.15.

**Comment by iMartyn** [comment]
Are we back to screaming into the void on this one?  I'd really like to see this one get in so I can at Kubeconf EU not be depressed when talking to people about contributing to k8s ;-)

**Comment by dims** [comment]
/hold cancel
/approve

(hold was about 1.14 release, now that's over unholding)

File: pkg/cloudprovider/providers/aws/aws.go [hunk 1]
```diff
@@ -3139,7 +3139,7 @@ func buildListener(port v1.ServicePort, annotations map[string]string, sslPorts
 	listener.InstancePort = &instancePort
 	listener.LoadBalancerPort = &loadBalancerPort
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: I don't think LoadBalancerPort is changed from the service port, so I don't think a service that is configured in k8s with port 80 is still going to listen on port 80?  Unless one of the other options overrides things?

File: pkg/cloudprovider/providers/aws/aws.go [hunk 2]
```diff
@@ -3139,7 +3139,7 @@ func buildListener(port v1.ServicePort, annotations map[string]string, sslPorts
...
 	certID := annotations[ServiceAnnotationLoadBalancerCertificate]
-	if certID != "" && (sslPorts == nil || sslPorts.numbers.Has(loadBalancerPort) || sslPorts.names.Has(portName)) {
+	if certID != "" && (sslPorts == nil || sslPorts.numbers.Has(loadBalancerPort) || sslPorts.names.Has(portName) || len(sslPorts.numbers) > 0) {
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: I don't think this is right - today when I have a k8s service that listens on port 80 and on port 443, I can configure `load-balancer-ssl-ports=443`, and then ELB will do https only on port 443, port 80 will listen for http.  The current behaviour will then send non-TLS traffic onto port 443 which is at the very least unintuitive.

But if we add `len(sslPorts.numbers) > 0` here we will break port 80.

**Comment by justinsb** [comment]
Sorry - I don't think this does what we think it does, and I think it also breaks mixed TLS / non-TLS services.  I commented on the underlying issue with some thoughts on how we can address the issue.

/hold

**Comment by iMartyn** [comment]
Well then, that's that.  Had these concerns been properly explained a year and a half ago instead of being fought every step of the way with people refusing that a problem exists or it just being abandoned then even a year ago I might have been able to continue this contribution, now it needs a new champion.

**Comment by iMartyn** [comment]
@unguiculus this is the PR I was talking about the other night at kubecon, feel free to see if you fancy being the champion.

**Comment by mcrute** [comment]
@iMartyn sorry I've been slow on this. The testing has not been as straightforward as I'd hoped and I got pulled into some other stuff. This is still on my radar and I'm planning to get back to it soon. Although I'm completely happy to turn it over to someone else if they have the time to push it forward faster than I.

**Pull request closed without merging**